controllers/tf_controller.go

```diff
@@ -518,6 +533,12 @@ func (r *TerraformReconciler) SetupWithManager(mgr ctrl.Manager, maxConcurrentRe
 func (r *TerraformReconciler) checkDependencies(source sourcev1.Source, terraform infrav1.Terraform) error {
 	dependantFinalizer := infrav1.TFDependencyOfPrefix + terraform.GetName()
 	for _, d := range terraform.Spec.DependsOn {
+		if r.NoCrossNamespaceRefs && d.Namespace != terraform.GetNamespace() {
```

Comment:
I think you'll want this check to go after defaulting the namespace, since otherwise the second part will be true any time a dependsOn entry has an empty `.namespace` (which should always pass the check, since it denotes a resource in the same namespace).

Reply:
In fact, I think you're better off constructing `dName`, then defaulting the namespace, then checking if it violates NoCrossNamespaceRefs; that way it avoids mutating d, which makes me a little nervous.